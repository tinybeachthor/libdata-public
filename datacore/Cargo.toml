[package]
name = "datacore"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
random-access-storage = { path = "../random-access-storage" }
anyhow = "1.0.26"
futures-lite = "1.12.0"
byteorder = "1.3.4"
ed25519-dalek = "1.0.1"
blake3 = "1.3.1"
hex = "0.4"
rand = { version = "0.7.3", features = [ "std", "wasm-bindgen" ] }

[dev-dependencies]
random-access-memory = { path = "../random-access-memory" }
random-access-disk = { path = "../random-access-disk" }
quickcheck = "0.9.2"
insta = "1.8.0"
async-std = { version = "1.10.0", features = ["attributes"] }
crypto-hash = "0.3.4"
tempfile = "3.1.0"
criterion = { version = "0.3.4", features = [ "async_std" ] }

[[bench]]
name = "io"
harness = false
