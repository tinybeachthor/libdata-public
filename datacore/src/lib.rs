#![forbid(unsafe_code, bad_style, nonstandard_style, future_incompatible)]
#![forbid(rust_2018_idioms, rust_2021_compatibility)]
#![deny(missing_debug_implementations)]
#![deny(missing_docs)]
#![cfg_attr(test, allow(dead_code))]
#![cfg_attr(test, deny(warnings))]

//! ## Introduction
//! Datacore is a secure, append-only, single-writer log.
//! It is meant for sharing large datasets and streams of real time data.
//! The primary way to use this crate is through the [Core] struct.
//!
//! ## Example
//! ```rust
//! # use async_std::prelude::*;
//! # use futures_lite::future::FutureExt;
//!
//! use random_access_memory::RandomAccessMemory;
//! use datacore::{Core, generate_keypair};
//!
//! # fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
//! # async_std::task::block_on(async {
//! let keypair = generate_keypair();
//! let mut core = Core::new(
//!     RandomAccessMemory::new(1024),
//!     RandomAccessMemory::new(1024),
//!     RandomAccessMemory::new(1024),
//!     keypair.public, Some(keypair.secret)
//!     ).await?;
//!
//! core.append(b"hello", None).await?;
//! core.append(b"world", None).await?;
//!
//! assert_eq!(core.len(), 2);
//! assert_eq!(
//!     core.get(0).await?.unwrap().0,
//!     b"hello");
//! assert_eq!(
//!     core.get(1).await?.unwrap().0,
//!     b"world");
//! # Ok(())
//! # })
//! # }
//! ```

mod block;
mod store_data;
mod store_blocks;
mod store_state;
mod merkle_tree_stream;
mod keys;
mod hash;
mod merkle;
mod core;

pub use random_access_storage::RandomAccess;
pub use block::{Signature, BlockSignature, Block, SIGNATURE_LENGTH};
pub use keys::{
    Keypair, PublicKey, SecretKey,
    generate_keypair, sign, verify
};
pub use hash::Hash;
pub use merkle::{Merkle, Node, NodeTrait};
pub use self::core::{Core, MAX_CORE_LENGTH, MAX_BLOCK_SIZE};
