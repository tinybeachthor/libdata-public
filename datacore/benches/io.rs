use criterion::{black_box, criterion_group, criterion_main, Criterion};
use criterion::async_executor::AsyncStdExecutor;
use async_std::task::block_on;

use random_access_memory::RandomAccessMemory;
use datacore::{generate_keypair, Core};

type HomogenousCore<T> = Core<T, T, T>;
type MemoryCore = HomogenousCore<RandomAccessMemory>;

fn random_access_memory() -> RandomAccessMemory {
    RandomAccessMemory::new(1024)
}

fn init() -> MemoryCore {
    block_on(async move {
        let keypair = generate_keypair();
        Core::new(
            random_access_memory(),
            random_access_memory(),
            random_access_memory(),
            keypair.public, Some(keypair.secret))
            .await.unwrap()
    })
}

async fn hypercore_append(mut core: MemoryCore, blocks: u64) {
    for i in 0..blocks {
        core.append(&i.to_be_bytes(), None).await.unwrap();
    }
}

pub fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("append 1000 blocks", |b| {
        b.to_async(AsyncStdExecutor).iter(|| {
            let feed = init();
            hypercore_append(black_box(feed), black_box(1_000))
        })
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
