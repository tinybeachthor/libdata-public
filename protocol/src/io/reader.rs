use std::io::{Error, ErrorKind, Result};
use std::pin::Pin;
use std::time::Duration;
use std::task::{Context, Poll};
use std::future::Future;
use futures_lite::io::AsyncRead;
use futures_timer::Delay;

use crate::message::{Frame, FrameType};
use crate::noise::{Cipher, HandshakeResult};
use crate::MAX_MESSAGE_SIZE;

const READ_BUF_INITIAL_SIZE: usize = 1024 * 128;

#[derive(Debug)]
pub struct ReadState {
    /// The read buffer.
    buf: Vec<u8>,
    /// The start of the not-yet-processed byte range in the read buffer.
    start: usize,
    /// The end of the not-yet-processed byte range in the read buffer.
    end: usize,
    /// The logical state of the reading (either header or body).
    step: Step,
    /// The timeout after which the connection is closed.
    timeout: Option<Delay>,
    /// Timeout duration.
    timeout_duration: Option<Duration>,
    /// Optional encryption cipher.
    cipher: Option<Cipher>,
    /// The frame type to be passed to the decoder.
    frame_type: FrameType,
}

impl ReadState {
    pub fn new(timeout_ms: Option<u64>) -> Self {
        let timeout_duration = timeout_ms.map(Duration::from_millis);
        Self {
            buf: vec![0u8; READ_BUF_INITIAL_SIZE as usize],
            start: 0,
            end: 0,
            step: Step::Header,
            timeout: timeout_duration.map(Delay::new),
            timeout_duration,
            cipher: None,
            frame_type: FrameType::Raw,
        }
    }
}

#[derive(Debug)]
enum Step {
    Header,
    Body { header_len: usize, body_len: usize },
}

impl ReadState {
    pub fn upgrade_with_handshake(&mut self, handshake: &HandshakeResult) -> Result<()> {
        let mut cipher = Cipher::from_handshake_rx(handshake)?;
        cipher.apply(&mut self.buf[self.start..self.end]);
        self.cipher = Some(cipher);
        Ok(())
    }

    pub fn set_frame_type(&mut self, frame_type: FrameType) {
        self.frame_type = frame_type;
    }

    pub fn poll_reader<R>(
        &mut self,
        cx: &mut Context<'_>,
        mut reader: &mut R,
    ) -> Poll<Result<Frame>>
    where
        R: AsyncRead + Unpin,
    {
        loop {
            if let Some(result) = self.process() {
                return Poll::Ready(result);
            }

            let n = match Pin::new(&mut reader).poll_read(cx, &mut self.buf[self.end..]) {
                Poll::Ready(Ok(n)) if n > 0 => n,
                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
                // If the reader is pending, poll the timeout.
                Poll::Pending | Poll::Ready(Ok(_)) => {
                    // Return Pending if the timeout is pending, or an error if the
                    // timeout expired (i.e. returned Poll::Ready).
                    return match self.timeout.as_mut() {
                        None => Poll::Pending,
                        Some(mut timeout) => match Pin::new(&mut timeout).poll(cx) {
                            Poll::Pending => Poll::Pending,
                            Poll::Ready(_) => Poll::Ready(Err(
                                    Error::new(ErrorKind::TimedOut, "Remote timed out"))),
                        },
                    }
                }
            };

            let end = self.end + n;
            if let Some(ref mut cipher) = self.cipher {
                cipher.apply(&mut self.buf[self.end..end]);
            }
            self.end = end;

            // reset timeout
            match self.timeout_duration {
                None => None,
                Some(timeout_duration) =>
                    self.timeout.as_mut().map(|t| t.reset(timeout_duration)),
            };
        }
    }

    fn cycle_buf_if_needed(&mut self) {
        // TODO: It would be great if we wouldn't have to allocate here.
        if self.end == self.buf.len() {
            let temp = self.buf[self.start..self.end].to_vec();
            let len = temp.len();
            self.buf[..len].copy_from_slice(&temp[..]);
            self.end = len;
            self.start = 0;
        }
    }

    fn process(&mut self) -> Option<Result<Frame>> {
        if self.start == self.end {
            return None;
        }
        loop {
            match self.step {
                Step::Header => {
                    let mut body_len = 0;
                    let header_len = varinteger::decode(
                        &self.buf[self.start..self.end], &mut body_len);

                    let body_len = body_len as usize;
                    if body_len > MAX_MESSAGE_SIZE as usize {
                        return Some(Err(Error::new(
                            ErrorKind::InvalidData,
                            "Message length above max allowed size",
                        )));
                    }
                    self.step = Step::Body {
                        header_len,
                        body_len,
                    };
                }
                Step::Body {
                    header_len,
                    body_len,
                } => {
                    let message_len = header_len + body_len;
                    if message_len > self.buf.len() {
                        self.buf.resize(message_len, 0u8);
                    }
                    if (self.end - self.start) < message_len {
                        self.cycle_buf_if_needed();
                        return None;
                    } else {
                        let range = self.start + header_len..self.start + message_len;
                        let frame = Frame::decode(&self.buf[range], &self.frame_type);
                        self.start += message_len;
                        self.step = Step::Header;
                        return Some(frame);
                    }
                }
            }
        }
    }
}
