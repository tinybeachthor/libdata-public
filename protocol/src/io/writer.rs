use std::fmt;
use std::io::Result;
use std::pin::Pin;
use std::task::{Context, Poll};
use std::collections::VecDeque;
use futures_lite::{ready, AsyncWrite};

use crate::message::{EncodeError, Encoder, Frame};
use crate::noise::{Cipher, HandshakeResult};

const BUF_SIZE: usize = 1024 * 64;

#[derive(Debug)]
pub enum Step {
    Flushing,
    Writing,
    Processing,
}

pub struct WriteState {
    queue: VecDeque<Frame>,
    buf: Vec<u8>,
    current_frame: Option<Frame>,
    start: usize,
    end: usize,
    cipher: Option<Cipher>,
    step: Step,
}

impl fmt::Debug for WriteState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("WriteState")
            .field("queue (len)", &self.queue.len())
            .field("step", &self.step)
            .field("buf (len)", &self.buf.len())
            .field("current_frame", &self.current_frame)
            .field("start", &self.start)
            .field("end", &self.end)
            .field("cipher", &self.cipher.is_some())
            .finish()
    }
}

impl WriteState {
    pub fn new() -> Self {
        Self {
            queue: VecDeque::new(),
            buf: vec![0u8; BUF_SIZE],
            current_frame: None,
            start: 0,
            end: 0,
            cipher: None,
            step: Step::Processing,
        }
    }

    pub fn queue_frame<F>(&mut self, frame: F)
    where
        F: Into<Frame>,
    {
        self.queue.push_back(frame.into())
    }

    pub fn try_queue_direct<T: Encoder>(
        &mut self,
        frame: &T,
    ) -> std::result::Result<bool, EncodeError> {
        let len = frame.encoded_len();
        if self.buf.len() < len {
            self.buf.resize(len, 0u8);
        }
        if len > self.remaining() {
            return Ok(false);
        }
        let len = frame.encode(&mut self.buf[self.end..])?;
        self.advance(len);
        Ok(true)
    }

    pub fn can_park_frame(&self) -> bool {
        self.current_frame.is_none()
    }

    pub fn park_frame<F>(&mut self, frame: F)
    where
        F: Into<Frame>,
    {
        if self.current_frame.is_none() {
            self.current_frame = Some(frame.into())
        }
    }

    fn advance(&mut self, n: usize) {
        let end = self.end + n;
        if let Some(ref mut cipher) = self.cipher {
            cipher.apply(&mut self.buf[self.end..end]);
        }
        self.end = end;
    }

    pub fn upgrade_with_handshake(&mut self, handshake: &HandshakeResult) -> Result<()> {
        let cipher = Cipher::from_handshake_tx(handshake)?;
        self.cipher = Some(cipher);
        Ok(())
    }
    fn remaining(&self) -> usize {
        self.buf.len() - self.end
    }

    fn pending(&self) -> usize {
        self.end - self.start
    }

    pub fn poll_send<W>(&mut self, cx: &mut Context<'_>, mut writer: &mut W) -> Poll<Result<()>>
    where
        W: AsyncWrite + Unpin,
    {
        loop {
            self.step = match self.step {
                Step::Processing => {
                    if self.current_frame.is_none() && !self.queue.is_empty() {
                        self.current_frame = self.queue.pop_front();
                    }

                    if let Some(frame) = self.current_frame.take() {
                        if !self.try_queue_direct(&frame)? {
                            self.current_frame = Some(frame);
                        }
                    }
                    if self.pending() == 0 {
                        return Poll::Ready(Ok(()));
                    }
                    Step::Writing
                }
                Step::Writing => {
                    let n = ready!(
                        Pin::new(&mut writer).poll_write(cx, &self.buf[self.start..self.end])
                    )?;
                    self.start += n;
                    if self.start == self.end {
                        self.start = 0;
                        self.end = 0;
                    }
                    Step::Flushing
                }
                Step::Flushing => {
                    ready!(Pin::new(&mut writer).poll_flush(cx))?;
                    Step::Processing
                }
            }
        }
    }
}
