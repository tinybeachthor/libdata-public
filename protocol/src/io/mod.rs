mod reader;
mod writer;

use anyhow::{Result, anyhow};
use std::task::{Context, Poll};
use futures_lite::io::{AsyncRead, AsyncWrite};

use crate::Options;
use crate::message::{Frame, EncodeError};
use self::reader::ReadState;
use self::writer::WriteState;

#[derive(Debug)]
pub struct IO<T> {
    io: T,
    pub options: Options,
    pub read_state: ReadState,
    pub write_state: WriteState,
}

impl<T> IO<T>
where
    T: AsyncRead + AsyncWrite + Send + Unpin + 'static,
{
    pub fn new(io: T, options: Options) -> Self {
        let keepalive_ms = options.keepalive_ms;
        Self {
            io,
            options,
            read_state: ReadState::new(keepalive_ms),
            write_state: WriteState::new(),
        }
    }

    /// Poll for inbound messages and process them.
    pub fn poll_inbound_read(
        &mut self,
        cx: &mut Context<'_>,
        ) -> Result<Option<Frame>>
    {
        let msg = self.read_state.poll_reader(cx, &mut self.io);
        return match msg {
            Poll::Ready(Ok(message)) => Ok(Some(message)),
            Poll::Ready(Err(e)) => Err(anyhow!(e)),
            Poll::Pending => Ok(None),
        }
    }

    /// Poll for outbound messages and write them.
    pub fn poll_outbound_write(&mut self, cx: &mut Context<'_>) -> Result<()>
    {
        let poll = self.write_state.poll_send(cx, &mut self.io);
        if let Poll::Ready(Err(e)) = poll {
            return Err(anyhow!(e));
        }
        return Ok(());
    }

    pub fn queue_frame_direct(&mut self, body: Vec<u8>)
        -> std::result::Result<bool, EncodeError>
    {
        let frame = Frame::Raw(body);
        self.write_state.try_queue_direct(&frame)
    }
}
