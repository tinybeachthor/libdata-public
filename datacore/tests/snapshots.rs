mod common;
use common::random_access_disk;

use std::path::Path;
use std::fs::File;
use std::io::Read;
use async_std::test;
use tempfile;
use insta;

use datacore::{Core, Keypair};

fn read_bytes(dir: &Path, s: &str) -> Vec<u8> {
    let mut f = File::open(dir.join(s)).unwrap();
    let mut bytes = Vec::new();
    f.read_to_end(&mut bytes).unwrap();
    bytes
}

const KEYPAIR_BYTES: [u8; 64] = [
    86, 29, 202, 51, 72, 159, 192, 155, 76, 131, 249, 122, 241, 244, 9, 244,
    157, 139, 190, 59, 130, 201, 224, 241, 121, 161, 171, 30, 158, 108, 23, 0,
    184, 16, 141, 118, 116, 37, 127, 146, 105, 139, 107, 124, 101, 123, 166,
    152, 83, 209, 170, 236, 172, 23, 111, 253, 30, 197, 241, 83, 169, 233,
    237, 77
];

#[test]
pub async fn snapshots_append() {
    let dir = tempfile::tempdir().unwrap().into_path();
    let keypair = Keypair::from_bytes(&KEYPAIR_BYTES).unwrap();
    let mut core = Core::new(
        random_access_disk(dir.to_path_buf().join("data")).await,
        random_access_disk(dir.to_path_buf().join("blocks")).await,
        random_access_disk(dir.to_path_buf().join("merkle")).await,
        keypair.public, Some(keypair.secret))
        .await.unwrap();

    let data = b"abcdef";
    for &b in data {
        core.append(&[b], None).await.unwrap();
    }
    assert_eq!(core.len() as usize, data.len());
    assert_eq!(read_bytes(&dir, "data"), data);

    insta::assert_debug_snapshot!(read_bytes(&dir, "blocks"));
    insta::assert_debug_snapshot!(read_bytes(&dir, "merkle"));
}
