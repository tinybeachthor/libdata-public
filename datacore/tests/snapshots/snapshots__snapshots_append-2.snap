---
source: datacore/tests/snapshots.rs
expression: "read_bytes(&dir, \"merkle\")"

---
[
    2,
    0,
    0,
    0,
    3,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    4,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    177,
    135,
    148,
    251,
    72,
    93,
    217,
    138,
    227,
    33,
    169,
    216,
    163,
    232,
    210,
    253,
    33,
    118,
    61,
    99,
    214,
    28,
    173,
    124,
    112,
    208,
    160,
    64,
    229,
    112,
    83,
    15,
    9,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    2,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    217,
    55,
    171,
    98,
    50,
    38,
    81,
    183,
    237,
    91,
    44,
    176,
    189,
    91,
    37,
    218,
    215,
    77,
    95,
    76,
    111,
    130,
    45,
    53,
    39,
    192,
    116,
    102,
    17,
    254,
    192,
    40,
]
